store/sqlstore/channel_store.go

```diff
@@ -1469,3 +1471,61 @@ func (s SqlChannelStore) GetMembersByIds(channelId string, userIds []string) sto
 		}
 	})
 }
+
+// GetChannelsWithChannelContextPermission retrieves the channels for a given user with an associated channel-context of a permission.
+func (s SqlChannelStore) GetChannelsWithChannelContextPermission(userID string, permissionName string) store.StoreChannel {
+	var query string
+
+	if s.DriverName() == model.DATABASE_DRIVER_POSTGRES {
+		query = getChannelsWithChannelContextPermissionPostgresQuery(permissionName)
+	} else {
+		// TODO, Phase 3 Advanced Permissions: role count will need to be queried with something like `select count(*) from roles` because of custom roles.
+		query = getChannelsWithChannelContextPermissionMysqlQuery(permissionName, TOTAL_ROLE_COUNT)
+	}
+
+	return store.Do(func(result *store.StoreResult) {
+		var channels model.ChannelList
+		_, err := s.GetReplica().Select(&channels, query, map[string]interface{}{"UserId": userID})
+		if err != nil {
+			fmt.Printf("err: %v", err)
+			result.Err = model.NewAppError("SqlChannelStore.GetChannelsWithChannelContextPermission", "store.sql_channel.search.app_error", nil, "", http.StatusInternalServerError)
+		} else {
+			result.Data = &channels
+		}
+	})
+}
+
+func getChannelsWithChannelContextPermissionMysqlQuery(permissionName string, roleCount int) string {
+	query := `
+			select Channels.* from (
+			select
+			ChannelId, UserId, trim(substring_index(substring_index(B.Roles, ' ', ns.n), ' ', -1)) as rolename
+			from (
+			select 1 as n union all `
+
+	for i := 2; i < roleCount; i++ {
+		query += fmt.Sprintf("select %v union all\n", i)
+	}
+
+	query += `select ` + fmt.Sprintf("%v", roleCount) + `
+			) ns
+			inner join ChannelMembers B ON ns.n <= char_length(B.Roles) - char_length(replace(B.Roles, ' ', '')) + 1
+			) x
+			left join Channels on ChannelId = Channels.Id
+			left join Roles on Roles.name = x.rolename
+			where Roles.Permissions REGEXP '[[:<:]]` + permissionName + `[[:>:]]'
+			and UserId = :UserId`
```

Comment:
This is quite the query, mind giving a brief overview of what it's doing?

Reply:
@jwilander I think the easiest way to understand it is to look at the Postgres query:

```sql
SELECT channels.* FROM (
    SELECT userid, channelid, UNNEST(STRING_TO_ARRAY(roles, ' ')) AS rolename FROM channelmembers
) x
LEFT JOIN channels ON channelid = channels.id
LEFT JOIN roles ON roles.name = x.rolename
WHERE STRING_TO_ARRAY(roles.permissions, ' ') && '{read_channel}'::TEXT[]
AND userid = 'jwedp8tbubn68mzbf3ufw1xczo';
```

The goal is to retrieve a list of channels that "have" the given permission. These are the pseudo steps:
* Get all channelmembers with the userid 'jwedp8tbubn68mzbf3ufw1xczo'.
* Unnest the roles (because they're stored as a string array) on channelmembers. This basically gives you something like a row per item in the string array so we can join the roles on them.
* Join the channels and roles tables on the results.
* Return only the channels that have the role 'read_channel' in the permissions of the role.

The MySQL query does the same thing just much less elegantly because it doesn't have the ability to "unnest" so you have to the ugly `select [N] union all` for each item in the roles array.